Refactor and enhance Colorize

MakeNowJust opened the pull request:
- Refactor all codes.
  * Prefer to use enums instead of constants.
  * Separate `Object` and `Style`. `Style` represents a style on terminal purely, however `Object` has object content.
- Support 256 color and 32bit true color.
- Support TTY detection. Three policy, they looks like GNU coreutils option `--color WHEN`:
  * `Auto`: If given io is TTY, then output colorized text. (default)
  * `Always`: Output colorized text always.
  * `Never`: Not output colorized text.
- Add command line option for `--color` to specify colorize policy.
  * Add `color` option to Makefile.
- Comment for document more.
- Rewrite tests for this changes.
- More and more...

1/23 edited.

Sija commented on src/colorize/style.cr:
```diff
@@ -0,0 +1,72 @@
+  #
+  # It outputs escape sequence, then invoke the block. After all, it outputs reset escape sequence if needed.
+  #
+  # This method has a stack internally, so it keeps highlight if nested.
```
`# NOTE: This method ...`

Sija commented on src/colorize/style.cr:
```diff
@@ -0,0 +1,72 @@
+
+  # Highlight the content in the block with this style
+  #
+  # It outputs escape sequence, then invoke the block. After all, it outputs reset escape sequence if needed.
```
typo: `invoke` -> `invokes`

Sija commented on src/colorize/style.cr:
```diff
@@ -0,0 +1,72 @@
+
+  # Get escape sequence text.
+  #
+  # NOTE: Thie methods outputs escape sequence if *io* is not tty. This behavior is difference from other `Colorize` methods.
```
`difference` -> `different`

Sija replied:
typo: `Thie` -> `This`

Sija commented on src/colorize/style.cr:
```diff
@@ -0,0 +1,72 @@
+    Colorize.write_style self, io, force: true
+  end
+
+  private STACK = [] of Colorize::Style
```
Using `Deque` here might yield some perf optimization.

MakeNowJust commented:
@Sija Updated. Thanks!

MakeNowJust commented:
This is still WIP.

MakeNowJust commented:
Complete to refactor `Colorize`!

Breaking changes in this pull request:

  - Remove `inspect` for `Colorize::Object`, because it complicates to debug `Colorize::Object` itself. And, we can simply replace `with_color.surround { |io| obj.inspect io }` in many case.
  - Change `Colorize::Style#surround` semantics a bit (It is same as old `Colorize::Object#surround`.) Its name makes more sense. `push` is just internal thing, and it is not important for users.

Only!

@Sija @asterite Could you review it?

Sija commented on src/colorize/io.cr:
```diff
@@ -164,11 +165,17 @@ module Colorize
   end
 end
+
+class IO::FileDescriptor
```
Why not put it inside `IOExtension`?

MakeNowJust replied:
I think it is just a matter of personal preference. But original `colorize.cr` extends `Object` class outside `ObjectExtension`, I followed it.

MakeNowJust commented:
Please review it... I desire to merge it because I have colorized `PrettyPrint` patch which requires this pull request changes (below is screen shot of patched `p({...})` method output):

![screen shot of patched p](https://cloud.githubusercontent.com/assets/6679325/22455021/28277be2-e7cd-11e6-8df9-689c89359eeb.png)

Thanks!

MakeNowJust commented:
Rebased to resolve conflict.

straight-shoota commented on src/compiler/crystal/command/format.cr:
```diff
@@ -162,21 +169,21 @@ class Crystal::Command
         check_files << FormatResult.new(filename, FormatResult::Code::SYNTAX)
       else
-        STDOUT << "Syntax Error:".colorize(:yellow).toggle(@color) << " " << ex.message << " at " << filename << ":" << ex.line_number << ":" << ex.column_number << "\n"
+        STDOUT << "Syntax Error:".colorize(:yello) << " " << ex.message << " at " << filename << ":" << ex.line_number << ":" << ex.column_number << "\n"
```
typo: yellow

MakeNowJust replied:
Oh, thanks!! I'll fix it.

spalladino commented:
Thanks for the PR @MakeNowJust! Please in the future do open an issue to discuss the design before sending a PR. Though adding a feature such as 32bit colors is one thing, introducing a major refactor that includes breaking changes deserves to quick discussion before actually going into it. We'll take a look at the PR later nonetheless, since it introduces many cool new features, and can discuss such things here; but please consider opening a discussion before submitting a PR that introduces major changes.

MakeNowJust commented:
@spalladino Ok, I understand. Should I create an issue to explain about the problem of current `Colorize` implementation?

spalladino commented:
We can keep that discussion here. The idea behind opening an issue beforehand to discuss a new implementation is to avoid working on a new feature in case the base idea is rejected. Going back to your question, actually a few lines on what prompted the refactor would be really helpful.

MakeNowJust commented:
@spalladino  Note for Refactoring points:

  - [Original code uses many constants](https://github.com/crystal-lang/crystal/blob/master/src/colorize.cr#L105-L159). These can be classified to colors and mode, so I prefer to use `enum`s for these. I added two `enum`s named `ColorANSI` and `Mode`.
  - Original code is only one file. On refactoring, I felt `src/colorize.cr` is being large. I separated `src/colorize.cr` contents under to `src/colorize/` directory. It made another benefit. Now, we can use `Colorize` without defining `Object#colorize` and `with_color` method on the global.
  - Original `Colorize::Object` has two usages: colorizing given object by `#to_s` and colorizing block output by `#push`. But, the given object does not make sense when using `#push`. I think two usages should not be contained by the class `Colorize::Object`, so I created new class `Colorize::Style`. (Also I renamed `#push` to `#surround`. This reason was explained in https://github.com/crystal-lang/crystal/pull/3925#issuecomment-275852281)

ysbaddaden commented:
Can you write a before/after example, so we can visually see the difference in the API?

MakeNowJust commented:
I said in https://github.com/crystal-lang/crystal/pull/3925#issuecomment-275852281, the breaking change is a bit. And, I wrote many doc comments (1/3 of all diffs is doc comment in fact.) Could you look these at once?

ysbaddaden commented:
The good :+1:
- grouping constants into enums (I would name `ANSIColor` or just `Color`, though);
- separating files so we can use `Colorize` without leaking to `Object` and global;
-  renaming `#push` to `#surround`.

The bad :-1:
- `When`, `IOExtension`, `#colorize_when`, leaking into `IO::FileDescriptor`, and everything related...

It's a good idea to have an `IO::Colorize` wrapper that nicely streams escape codes, instead of allocating strings just to write them to the IO (pushing pressure on GC), but this shouldn't leak into an IO implementation in particular, nor leak at all.

I'm not sure I see a real usage for `When`; maybe just a mean to disable `IO::Colorize` so it acts as a null colorizer (useful when `--no-color` is passed) along with an `enabled?` or `disabled?` accessor, would be enough.

Overall, I feel like the implementation could be a little more straightforward and use less types. A struct with a few methods to add/remove modes and styles on strings, another one wrapping IO objects, some enums, and that's it.

ysbaddaden commented:
BTW: `Colorize` is a nice name, but quite generic. Maybe `ANSI` would be more specific, with `ANSI::Color`, `ANSI::Style`, and so on, but also less discoverable, unless you already know what [ANSI escape codes](https://en.wikipedia.org/wiki/ANSI_escape_code) are.

asterite commented:
@MakeNowJust I understand your points, but:

1. The constants are never used by a user. You usually do `colorize.red`, `colorize.fore(:green)`, etc.
2. One file: I think that's better (easier to understand, easier to read, easier to change) than having 8 files for the code.
3. The original file defines one type, now we have like 13 types. I understand this is common in Ruby (let's objectify everything) but it's not my preferred style, and it's definitely not the style that the std follows.

I think this PR introduces a lot of unnecessary complexities. I'm also not sure about modifying IO. The logic of whether color should be used if IO is tty or not doesn't belong here. Can't we just add `colorize.fore(r: 1, g: 2, b: 3)` and `.fore(rgb: "#abcdef")` and keep the representation internal?

MakeNowJust commented:
@ysbaddaden 

> The bad 👎
>
>    - When, IOExtension, #colorize_when, leaking into IO::FileDescriptor, and everything related...

I believe my implementation is almost perfect solution for auto TTY detecting. But most people don't want perfect implemetation for this problem, and this problem is *too hard* for human brain. I'll find simpler solution.

> BTW: Colorize is a nice name, but quite generic. Maybe ANSI would be more specific, with ANSI::Color, ANSI::Style, and so on, but also less discoverable, unless you already know what ANSI escape codes are.

It sounds good. In fact, [Elixier's same thing](https://hexdocs.pm/elixir/IO.ANSI.html) is named `ANSI`.

@asterite 

> 1. The constants are never used by a user. You usually do colorize.red, colorize.fore(:green), etc.

I said just "refactoring", this fix is for maintenancibility.

> 3. The original file defines one type, now we have like 13 types. I understand this is common in Ruby (let's objectify everything) but it's not my preferred style, and it's definitely not the style that the std follows.

I think `Object` and `Style` should be separated at least, and `StyleBuilder` mixin is needed for them. This is just my preference, I hope you to understand.

MakeNowJust commented:
@ysbaddaden `IO::Colorize` is good idea. But this requires to create a new file `src/io/colorize.cr` and `Colorize` logic will be here. I think it makes `Colorize` implementation unclear.

ozra commented:
ANSI is a bad choice, imo. Colorize sounds great. That's what it does! It's rather obvious that it's performed using the ANSI standard for the escape codes?

In contrast, what's an "ANSI"-module? Does it contain an ANSI C-parser? So we can have a separate K&R?

_Ok, I'm bitching a bit_, but when I hear ANSI the first thing I think of is actually C, but I guess I'm quiet in the loner-room there. 

My first thought would definitely be to search for "text colorizer", next "ansi _color codes_" or something similar.

Discoverability as @ysbaddaden mentioned - that's a good thing!

No strong opinions really - just stumbled across - so, there, 2 cents ;-)

sdogruyol commented:
This is a good PR waiting to be merged / reviewed 💯

Sija commented:
bump 🚧

sdogruyol commented:
Another bump for this PR :+1: @MakeNowJust

asterite commented:
The code is too complex for my taste. Sorry, but I can't approve it like that.

j8r commented on src/colorize/io.cr:
```diff
@@ -0,0 +1,181 @@
+    def colorize_write(style, reset = false)
+      return false if style.all_default? && !reset
+
+      self << "\e["
```
Please use the POSIX compliant escape sequence `\33` (or `\033` but it seems to do the same) instead of the `bash` one `\e` (doesn't work in `dash` for example)

MakeNowJust replied:
Why?

j8r replied:
Oh ok nevermind, there is not need to change so.

j8r commented on src/colorize/io.cr:
```diff
@@ -0,0 +1,181 @@
+
+    # :nodoc:
+    def colorize_reset
+      self << "\e[0m"
```
Ditto

straight-shoota commented:
This needs a rebase now.

straight-shoota commented:
Closing stale PR. There has been no progress in over two years and there are a lot of merge conflicts by now.
It seems this change was too big and resulted in too much complexity to be comfortably approved.

Refactoring and enhancements to ` Colorize` are still welcome, but it should probably start anew in smaller bites. I think some of the proposed changes have already been applied or don't apply anymore.
